### PR TITLE
[optim] Improve adadelta foreach, group tensors to maximize fast path

### PR DESCRIPTION
**Pull request by janeyx99**
Old behavior would have adadelta foreach sending tensors to the slow path if they were not all the same dtype nor on the same device.

This PR adds grouping for adadelta optimizer so that it would run foreach in batches, allowing more users to benefit from foreach perf.

Of course, we should ensure that the new implementation works, so there are new tests to ensure this behavior is not broken.

### CODE REVIEW

File: .jenkins/pytorch/multigpu-test.sh [hunk 1]
```diff
@@ -45,4 +45,5 @@ time python test/run_test.py --verbose -i distributed/_shard/test_partial_tensor
 time python test/run_test.py --verbose -i distributed/_shard/test_replicated_tensor
 # Other tests
 time python test/run_test.py --verbose -i test_cuda_primary_ctx
+time python test/run_test.py --verbose -i test_optim -- -k optimizers_with_varying_tensors
```
Comment (janeyx99) [inline .jenkins/pytorch/multigpu-test.sh:48]: @pytorch/pytorch-dev-infra to make sure this is okay. The total time it would add to multigpu is about 13 seconds.

File: test/test_optim.py [hunk 2]
```diff
@@ -750,6 +818,16 @@ def test_multi_tensor_optimizers(self):
         ]
         self._test_derived_optimizers(optimizer_pairs_with_flags, "foreach")
 
+    @unittest.skipIf(not TEST_MULTIGPU, "only one GPU detected")
+    def test_multi_tensor_optimizers_with_varying_tensors(self):
+        optimizer_pairs_with_flags = [
+            (optim.Adadelta, dict(weight_decay=0)),
+            (optim.Adadelta, dict(weight_decay=1)),
+            (optim.Adagrad, dict(weight_decay=0)),
+            (optim.Adagrad, dict(weight_decay=1)),
```
Comment (albanD) [inline test/test_optim.py:827]: Why are we not testing all the optimizers here?

**Review by albanD — COMMENTED** [review]
Looks good, just the seed needs to be change, the rest are nits

File: test/test_optim.py (hunk 2 above)
Comment (janeyx99) [reply test/test_optim.py:827]: Cuz the majority of them just fall back to the slow implementation anyway and it's more interesting to add these tests when I change their implementations to use the group by util.

File: test/test_optim.py [hunk 3]
```diff
@@ -645,6 +646,73 @@ def test_sgd_complex(self):
...
+        for optimizer_constructor, kwargs in optimizer_with_kwargs:
+            res, state = [], []
+            for enabled in (False, True):
+                kwargs_clone = deepcopy(kwargs)
+                kwargs_clone[kwarg] = enabled
+
+                params_clone = []
+                for p in params:
+                    p_clone = torch.clone(p).detach()
+                    p_clone.requires_grad = True
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply test/test_optim.py:684]: TIL!

File: test/test_optim.py [hunk 4]
```diff
@@ -645,6 +646,73 @@ def test_sgd_complex(self):
...
+
+        for optimizer_constructor, kwargs in optimizer_with_kwargs:
+            res, state = [], []
+            for enabled in (False, True):
+                kwargs_clone = deepcopy(kwargs)
+                kwargs_clone[kwarg] = enabled
+
+                params_clone = []
+                for p in params:
+                    p_clone = torch.clone(p).detach()
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply test/test_optim.py:683]: Another TIL!

File: test/test_optim.py [hunk 5]
```diff
@@ -645,6 +646,73 @@ def test_sgd_complex(self):
                 )
             )
 
+
+    def _test_derived_optimizers_varying_tensors(self, optimizer_with_kwargs, kwarg):
+        if not torch.cuda.is_available():
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply test/test_optim.py:651]: Yes, but for people running it locally...

File: test/test_optim.py (hunk 2 above)
Comment (janeyx99) [reply test/test_optim.py:856]: I added the majority anyway. Will add NAdam when the time comes.

File: test/test_optim.py [hunk 6]
```diff
@@ -645,6 +646,75 @@ def test_sgd_complex(self):
...
+        params = [
+            torch.rand(2, 3, dtype=torch.float64, device='cuda:0', requires_grad=True),
+            torch.rand(2, 3, dtype=torch.float32, device='cuda:0', requires_grad=True),
+            torch.rand(2, 3, dtype=torch.float16, device='cuda:0', requires_grad=True),
+            torch.rand(2, 3, dtype=torch.bfloat16, device='cuda:0', requires_grad=True),
+            torch.rand(2, 3, dtype=torch.float64, device='cuda:1', requires_grad=True),
+            torch.rand(2, 3, dtype=torch.float32, device='cuda:1', requires_grad=True),
+            torch.rand(2, 3, dtype=torch.float16, device='cuda:1', requires_grad=True),
+            torch.rand(2, 3, dtype=torch.bfloat16, device='cuda:1', requires_grad=True),
+            torch.randint(1024, (2, 3), dtype=torch.int64, device='cuda:1', requires_grad=False),
```
Comment (janeyx99) [inline test/test_optim.py:669]: New change from before: I threw in an int tensor to stress test a little more.

**Review by albanD — APPROVED** [review]
SGTM
